### PR TITLE
[Fix #437] Fix a false positive for `Performance/ChainArrayAllocation`

### LINKED ISSUE #437: Incorrect recommendation suggested by Performance/ChainArrayAllocation Cop
## Expected behavior

RuboCop suggests valid corrections on arrays.

## Actual behavior

Identifies multiple uses of ActiveRecord query `select` [method](https://apidock.com/rails/ActiveRecord/QueryMethods/select)
While calling `select` multiple times on the same relation may technically be redundant, this is a completely distinct method from `Array.select`, and thus suggesting to use `select!` (which does not exist), should not happen.

## Steps to reproduce the problem

Run rubocop on this file with `Performance/ChainArrayAllocation` enabled

models/test.rb
```
class Test < ActiveRecord::Base

  scope :test_scope, ->  {
    select(:field_one, :field_two).
    select("COUNT(DISTINCT(tests.field_three)) AS num").
    where("tests.field_one IS NOT NULL").
    group(:field_two)
  }

  def self.something
    Test.all
        .select(:field_one, :field_two)
        .select("COUNT(DISTINCT(tests.field_three)) AS num")
        .where("tests.field_one IS NOT NULL")
        .group(:field_two)
  end

end
```

## RuboCop version

Found on version
```
1.31.0 (using Parser 3.3.0.5, rubocop-ast 1.30.0, running on ruby 2.7.8 x86_64-darwin21)
  - rubocop-performance 1.19.1
  - rubocop-rails 2.15.2
```

Still occurs on latest
```
1.60.2 (using Parser 3.3.0.5, rubocop-ast 1.30.0, running on ruby 2.7.8) [x86_64-darwin21]
  - rubocop-performance 1.20.2
  - rubocop-rails 2.23.1
```

----------

Perhaps this is why this cop is disabled by default, because it has these are unsolvable false positives?

### PR DESCRIPTION
**Pull request by koic**
Fixes #437.

This PR fixes a false positive for `Performance/ChainArrayAllocation` when using `select` with block argument after `select`.

-----------------

Before submitting the PR make sure the following are checked:

* [x] The PR relates to *only* one subject with a clear title and description in grammatically correct, complete sentences.
* [x] Wrote [good commit messages][1].
* [x] Commit message starts with `[Fix #issue-number]` (if the related issue exists).
* [x] Feature branch is up-to-date with `master` (if not - rebase it).
* [x] Squashed related commits together.
* [x] Added tests.
* [x] Ran `bundle exec rake default`. It executes all tests and runs RuboCop on its own code.
* [x] Added an entry (file) to the [changelog folder](https://github.com/rubocop/rubocop-performance/blob/master/changelog/) named `{change_type}_{change_description}.md` if the new code introduces user-observable changes. See [changelog entry format](https://github.com/rubocop/rubocop/blob/master/CONTRIBUTING.md#changelog-entry-format) for details.

[1]: https://chris.beams.io/posts/git-commit/

### CODE REVIEW

File: spec/rubocop/cop/performance/chain_array_allocation_spec.rb [hunk 1]
```diff
@@ -53,4 +53,21 @@
...
   end
+
+  describe 'when using `select` with block argument after `select`' do
+    it 'registers an offense' do
+      expect_offense(<<~RUBY)
+        model.select(:foo, :bar).select { |item| item.do_something }
+                                ^^^^^^^ Use unchained `select` and `select!` (followed by `return array` if required) instead of chaining `select...select`.
+      RUBY
+    end
+  end
```
Comment (matthewhively) [inline spec/rubocop/cop/performance/chain_array_allocation_spec.rb:64]: Forgive my possible stupidity (I'm not terribly familiar with how the tests in rubocop work)
But wouldn't this still be wrong?
Wouldn't it flag `select` (QueryMethods#select) followed by a `select` (enumerable).
Which isn't a double select that requires optimization?

You would however want to flag the 3rd select here right?
`model.select(:foo, :bar).select { |item| item.do_something }.select { |item| item.do_something_else }`

Comment (koic) [reply spec/rubocop/cop/performance/chain_array_allocation_spec.rb:64]: I'm not entirely sure, this patch works as follows:

```ruby
it 'registers an offense' do
  expect_offense(<<~RUBY)
    model.select(:foo, :bar).select { |item| item.do_something }.select { |item| item.do_something_else }
                            ^^^^^^^ Use unchained `select` and `select!` (followed by `return array` if required) instead of chaining `select...select`.
                                                                ^^^^^^^ Use unchained `select` and `select!` (followed by `return array` if required) instead of chaining `select...select`.
  RUBY
end
```

So, this is a patch that distinguishes between `select(:arg)` and `select { |x| x }`. What potential issues might arise from this?

Comment (matthewhively) [reply spec/rubocop/cop/performance/chain_array_allocation_spec.rb:64]: So, let me rephrase
with this code 
```
expect_offense(<<~RUBY)
        model.select(:foo, :bar).select { |item| item.do_something }
                                ^^^^^^^ Use unchained `select` and `select!` (followed by `return array` if required) instead of chaining `select...select`.
RUBY
```
What should the suggested corrected code look like?

Comment (koic) [reply spec/rubocop/cop/performance/chain_array_allocation_spec.rb]: Ah, I understand. This should definitely be recognized as a false positive and I've fixed it. Thank you!
